File: server/Data/Entities/OngoingWarrantPosition.cs
using System;
using System.ComponentModel.DataAnnotations;
using SolidTradeServer.Data.Entities.Common;
using SolidTradeServer.Data.Models.Enums;

namespace SolidTradeServer.Data.Entities
{
    public class OngoingWarrantPosition : BaseEntity
    {
        [Required] 
        public DateTimeOffset GoodUntil { get; set; }
        
        [Required]
        public Portfolio Portfolio { get; set; }
        
        [Required]
        public WarrantPosition CurrentWarrantPosition { get; set; }
        
        [Required]
        [MaxLength(12)]
        public string Isin { get; set; }
        
        [Required]
        public EnterOrExitPositionType Type { get; set; }
        
        [Required]
        [Range(0.00010, int.MaxValue)]
        public decimal Price { get; set; }
         
        [Required]
        [Range(1, int.MaxValue)]
        public decimal NumberOfShares  { get; set; }
    }
}

File: client/lib/data/entities/ongoing_warrant_position.dart
import 'package:enum_to_string/enum_to_string.dart';
import 'package:solidtrade/data/enums/enter_or_exit_position_type.dart';
import 'package:solidtrade/data/entities/base/base_entity.dart';

class OngoingWarrantPosition implements IBaseEntity {
  @override
  final int id;
  @override
  final DateTime createdAt;
  @override
  final DateTime updatedAt;

  final String isin;
  final double price;
  final double numberOfShares;
  final DateTime goodUntil;
  final EnterOrExitPositionType type;

  const OngoingWarrantPosition({
    required this.id,
    required this.createdAt,
    required this.updatedAt,
    required this.isin,
    required this.price,
    required this.numberOfShares,
    required this.goodUntil,
    required this.type,
  });

  factory OngoingWarrantPosition.fromJson(Map<String, dynamic> json) {
    return OngoingWarrantPosition(
      id: json["id"],
      createdAt: DateTime.parse(json["createdAt"]),
      updatedAt: DateTime.parse(json["updatedAt"]),
      goodUntil: DateTime.parse(json["goodUntil"]),
      type: EnumToString.fromString(EnterOrExitPositionType.values, json["type"])!,
      isin: json["isin"],
      price: json["price"],
      numberOfShares: json["numberOfShares"],
    );
  }
}
